dff/utils/benchmark/context_storage.py

```diff
@@ -0,0 +1,585 @@
+        BenchmarkCase.set_average_results(benchmark)
+        return benchmark
+
+
+def save_results_to_file(
+    benchmark_cases: tp.List[BenchmarkCase],
+    file: tp.Union[str, pathlib.Path],
+    name: str,
+    description: str,
+    exist_ok: bool = False,
+):
+    """
+    Benchmark all `benchmark_cases` and save results to a file.
+
+    Result are saved in json format with this schema:
+    `utils/db_benchmark/benchmark_schema.json <../../../utils/db_benchmark/benchmark_schema.json>`_.
+
+    Files created by this function cen be viewed either by using :py:func:`~.report` or
+    streamlit app located in the utils directory:
+    `utils/db_benchmark/benchmark_streamlit.py <../../../utils/db_benchmark/benchmark_streamlit.py>`_.
+
+    :param benchmark_cases: A list of benchmark cases that specify benchmarks.
+    :param file: File to save results to.
+    :param name: Name of the benchmark set.
+    :param description: Description of the benchmark set.
+    :param exist_ok: Whether to continue if the file already exists.
+    """
+    with open(file, "w" if exist_ok else "x", encoding="utf-8") as fd:
+        uuid = str(uuid4())
+        result: tp.Dict[str, tp.Any] = {
+            "name": name,
+            "description": description,
+            "uuid": uuid,
+            "benchmarks": [],
+        }
+        cases = tqdm(benchmark_cases, leave=False)
+        for case in cases:
+            cases.set_description(f"Benchmarking: {case.name}")
+            result["benchmarks"].append({**case.dict(), "sizes": case.benchmark_config.sizes(), **case.run()})
+
+        json.dump(result, fd)
+
+
+def benchmark_all(
+    file: tp.Union[str, pathlib.Path],
+    name: str,
+    description: str,
+    db_uri: str,
+    benchmark_configs: tp.Dict[str, BenchmarkConfig],
+    exist_ok: bool = False,
+):
+    """
+    A wrapper for :py:func:`~.save_results_to_file`.
+
+    Generates `benchmark_cases` from `db_uri` and `benchmark_configs`:
+    `db_uri` is used to initialize :py:class:`~.DBFactory` instance
+    which is then used along with `benchmark_configs` to initialize :py:class:`~.BenchmarkCase` instances.
+
+    :param file: File to save results to.
+    :param name: Name of the benchmark set.
+    :param description: Description of the benchmark set. The same description is used for benchmark cases.
+    :param db_uri: URI of the database to benchmark
+    :param benchmark_configs: Mapping from case names to configs.
+    :param exist_ok: Whether to continue if the file already exists.
+    """
+    save_results_to_file(
+        [
+            BenchmarkCase(
+                name=case_name,
+                description=description,
+                db_factory=DBFactory(uri=db_uri),
+                benchmark_config=benchmark_config,
+            )
+            for case_name, benchmark_config in benchmark_configs.items()
+        ],
+        file,
+        name,
+        description,
+        exist_ok=exist_ok,
+    )
+
+
+def report(
```

Comment:
It seems to me (correct me if I'm wrong) that this file contains both *benchmarking* context storages and *retrieving* benchmark info (from files e.g.).
I consider these two to be completely different processes that might well be separated into two different source files.